File: app/src/main/java/com/hao/easy/base/BaseFragment.java
package com.hao.easy.base;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * @author Yang Shihao
 * @date 2018/10/1
 */
public abstract class BaseFragment extends Fragment {

    protected View rootView;
    protected Activity activity;
    private Unbinder unbinder;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(getLayoutId(), container, false);
        unbinder = ButterKnife.bind(this, rootView);
        activity = getActivity();
        return rootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initView();
        initData();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if(unbinder!= null){
            unbinder.unbind();
        }
    }

    protected abstract @LayoutRes
    int getLayoutId();

    protected abstract void initView();

    protected abstract void initData();
}

File: mvvm_base/src/main/java/com/hao/easy/mvvm/base/ui/BaseFragment.kt
package com.hao.easy.mvvm.base.ui

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.hao.easy.mvvm.base.di.android.DIFragment
import dagger.android.support.DaggerFragment

/**
 * @author Yang Shihao
 * @date 2018/11/18
 */
abstract class BaseFragment : DaggerFragment() {

    private lateinit var fragmentRootView: View
    private var isCreated = false
    private var isVisibleToUser = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        fragmentRootView = inflater.inflate(getLayoutId(), container, false)
        return fragmentRootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        onInit()
        initInject()
        initView()
        lodData()
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            this.isVisibleToUser = true
            lazyLoad()
        }
    }

    open fun onInit() {

    }

    private fun lodData() {
        if (isLazy()) {
            isCreated = true
            lazyLoad()
        } else {
            initData()
        }
    }

    private fun lazyLoad() {
        if (isCreated && isVisibleToUser) {
            initData()
            isCreated = false
        }
    }

    fun <T : View> f(id: Int): T? {
        return fragmentRootView?.findViewById(id)
    }

    @LayoutRes
    abstract fun getLayoutId(): Int

    open fun initInject() {

    }

    open fun initView() {

    }

    open fun initData() {

    }

    open fun isLazy() = false
}